share/rocm/cmake/ROCMSphinxDoc.cmake

```diff
@@ -1,63 +1,89 @@
 # ######################################################################################################################
-# Copyright (C) 2021 Advanced Micro Devices, Inc.
+# Copyright (C) 2023 Advanced Micro Devices, Inc.
 # ######################################################################################################################
 
+include_guard(GLOBAL)
+
 include(CMakeParseArguments)
 include(ROCMDocs)
 
 find_program(
     SPHINX_EXECUTABLE
     NAMES sphinx-build
-    HINTS $ENV{SPHINX_DIR}
+    HINTS "$ENV{SPHINX_DIR}"
     PATH_SUFFIXES bin
-    DOC "Sphinx documentation generator")
-
-mark_as_advanced(SPHINX_EXECUTABLE)
+    DOC "Sphinx documentation generator"
+    REQUIRED)
 
-set(BINARY_BUILD_DIR "${CMAKE_CURRENT_BINARY_DIR}/sphinx/_build")
+find_program(
+    DOXYGEN_EXECUTABLE
+    NAMES doxygen
+    HINTS "$ENV{DOXYGEN_DIR}"
+    PATH_SUFFIXES bin
+    DOC "Doxygen documentation generator")
 
-# Sphinx cache with pickled ReST documents
-set(SPHINX_CACHE_DIR "${CMAKE_CURRENT_BINARY_DIR}/sphinx/_doctrees")
+mark_as_advanced(SPHINX_EXECUTABLE)
 
-# HTML output directory
-set(SPHINX_DEFAULT_HTML_DIR "${CMAKE_CURRENT_BINARY_DIR}/sphinx/html")
 function(rocm_add_sphinx_doc SRC_DIR)
-    set(options)
+    set(options USES_DOXYGEN)
     set(oneValueArgs BUILDER OUTPUT_DIR)
     set(multiValueArgs DEPENDS VARS TEMPLATE_VARS)
 
     cmake_parse_arguments(PARSE "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
 
-    string(TOUPPER ${PARSE_BUILDER} BUILDER)
+    get_filename_component(SRC_DIR "${SRC_DIR}" ABSOLUTE BASE_DIR "${CMAKE_CURRENT_SOURCE_DIR}")
+    if(NOT EXISTS "${SRC_DIR}/conf.py")
+        message(FATAL_ERROR "rocm_add_sphinx_doc cannot find ${SRC_DIR}/conf.py")
+    endif()
 
-    set(ADDITIONAL_ARGS)
-    foreach(VAR ${PARSE_VARS})
-        list(APPEND ADDITIONAL_ARGS -D ${VAR})
+    if(PARSE_BUILDER)
+        string(TOUPPER ${PARSE_BUILDER} BUILDER)
+    else()
+        message(FATAL_ERROR "rocm_add_sphinx_doc requires providing a BUILDER to use.")
+    endif()
+
+    if(PARSE_OUTPUT_DIR)
+        get_filename_component(OUTPUT_DIR "${PARSE_OUTPUT_DIR}" ABSOLUTE BASE_DIR "${CMAKE_CURRENT_BINARY_DIR}")
+    else()
+        set(OUTPUT_DIR "${CMAKE_CURRENT_BINARY_DIR}/sphinx/${PARSE_BUILDER}")
+    endif()
```

Comment:
Maybe worth noting. I saw only 2 projects use this module. Hoping that will change in the future.

The cache variables in the long run have to go. While using `ExternalProject` excessively to drive the docs super build such that we could have multiple instances of these cache vars (one per project), the end game IMHO should be getting ROCm projects to a state where they are `add_subdirectory` friendly, at least among themselves and the entire thing (minus LLVM) can be built using one ninja process. If that's not something we're aspiring towards, I could restore the cache variables to how they were, but it will cause conflicts as soon as two documented projects share a build.

Reply:
I dont know how much this is used, but it is nice to be able to set the `OUTPUT_DIR` without needing to modify the source files. Maybe you could just do `if(DEFINED)` check and then use the variable. This shouldn't cause problems for `add_subdirectory`.

Reply:
Please take a look at the current solution driven by `ROCM_CMAKE_DOCS_DIR`. This was the closest to replicating a possibility of specifying externally an output directory while also keeping it idiomatic CMake (as far as command wrappers can be idiomatic). It tries to mimic globally defined variables that target-props take their default from, such as `CMAKE_RUNTIME_OUTPUT_DIRECTORY`.